src/serial/RS232Net.cc

```diff
@@ -0,0 +1,446 @@
+//      needed as IPv6 addresses themselves already contain colons,
+//      and it would be impossible to clearly distinguish between
+//      an IPv6 address, and an IPv6 address where a port has been
+//      specified. This format is a common one.
+//
+//
+bool RS232Net::network_address_generate_sockaddr(const char * address_string)
+{
+	struct addrinfo hints,*res;
+
+	memset(&hints,0,sizeof(hints));
+
+	/* preset the socket address */
+	memset(&socket_address.address, 0, sizeof socket_address.address);
+	socket_address.protocol = IPPROTO_TCP;
+
+	if (!address_string || address_string[0] == 0) {
+		/* there was no address given, do not try to process it. */
+		return false;
+	}
+
+	const char * port_part = NULL;
+	char * address_part = strdup(address_string);
+	/* try to find out if a port has been specified */
+	port_part = strchr(address_string, ']');
+		
+	if (port_part) {
+		if ((address_string[0] == '[') && (port_part[1] == ':')) {
+			/* [IPv6]:port format*/
+			char * p;
+
+			++port_part;
+			p = strdup(address_string+1);
+			p[port_part - (address_string+2)] = 0;
+			free(address_part);
+			address_part = p;
+			++port_part;
+			hints.ai_family = AF_INET6;
+			socket_address.domain = PF_INET6;
+			socket_address.len = sizeof socket_address.address.ipv6;
+			socket_address.address.ipv6.sin6_family = AF_INET6;
+			socket_address.address.ipv6.sin6_port = 0;
+			socket_address.address.ipv6.sin6_addr = in6addr_any;
+		} else {
+			/* malformed address, do not try to process it. */
+			free(address_part);
+			return false;
+		}
+	} else {
+		/* count number of colons */
+		int i = 0;
+		const char *pch=strchr(address_string,':');
+		while (pch!=NULL) {
+			i++;
+			pch=strchr(pch+1,':');
+		}
+		if (i == 1) {
+			/* either domainname:port or ipv4:port */
+			port_part = strchr(address_string,':');
+			char * p;
+
+			p = strdup(address_string);
+			p[port_part - address_string] = 0;
+			free(address_part);
+			address_part = p;
+			++port_part;
+			hints.ai_family = AF_INET;
+			socket_address.domain = PF_INET;
+			socket_address.len = sizeof socket_address.address.ipv4;
+			socket_address.address.ipv4.sin_family = AF_INET;
+			socket_address.address.ipv4.sin_port = 0;
+			socket_address.address.ipv4.sin_addr.s_addr = INADDR_ANY;
+		} else if (i > 1) {
+			/* ipv6 address */
+
+			hints.ai_family = AF_INET6;
+			socket_address.domain = PF_INET6;
+			socket_address.len = sizeof socket_address.address.ipv6;
+			socket_address.address.ipv6.sin6_family = AF_INET6;
+			socket_address.address.ipv6.sin6_port = 0;
+			socket_address.address.ipv6.sin6_addr = in6addr_any;
+		}
+	}
+
+	hints.ai_socktype = SOCK_STREAM;
+	hints.ai_protocol = IPPROTO_TCP;
+
+	int result = false;
+
+	if ( getaddrinfo(address_part,port_part,&hints,&res) == 0) {
+		memcpy(&socket_address.address, res->ai_addr, res->ai_addrlen);
+		result = true;
+	}
+	freeaddrinfo(res);
+
+	free(address_part);
+	return result;
+}
```

Comment:
String manipulation isn't easily, especially in C.
Using C++ combined with some extra openMSX utility functions allows to simplify this.
I think the following should be equivalent. This replaces `network_address_generate()` and `network_address_generate_sockaddr()`.
Disclaimer: I have NOT tested this.

    bool RS232Net::network_address_generate()
    {
            std::string_view address = rs232NetAddressSetting.getString();
            if (address.empty()) {
                    // there was no address given, do not try to process it.
                    return false;
            }
    
            // preset the socket address
            memset(&socket_address.address, 0, sizeof(socket_address.address));
            socket_address.protocol = IPPROTO_TCP;
    
            struct addrinfo hints;
            memset(&hints, 0, sizeof(hints));
            hints.ai_socktype = SOCK_STREAM;
            hints.ai_protocol = IPPROTO_TCP;
    
            auto setIPv4 = [&] {
                    hints.ai_family = AF_INET;
                    socket_address.domain = PF_INET;
                    socket_address.len = sizeof(socket_address.address.ipv4);
                    socket_address.address.ipv4.sin_family = AF_INET;
                    socket_address.address.ipv4.sin_port = 0;
                    socket_address.address.ipv4.sin_addr.s_addr = INADDR_ANY;
            };
            auto setIPv6 = [&] {
                    hints.ai_family = AF_INET6;
                    socket_address.domain = PF_INET6;
                    socket_address.len = sizeof(socket_address.address.ipv6);
                    socket_address.address.ipv6.sin6_family = AF_INET6;
                    socket_address.address.ipv6.sin6_port = 0;
                    socket_address.address.ipv6.sin6_addr = in6addr_any;
            };
    
            // Parse 'address', fills in 'addressPart' and 'portPart'
            std::string addressPart;
            std::string portPart;
            auto [ipv6_address_part, ipv6_port_part] = StringOp::splitOnFirst(address, ']');
            if (!ipv6_port_part.empty()) { // there was a ']' character (and it's already dropped)
                    // Try to parse as: "[IPv6]:port"
                    if (!ipv6_address_part.starts_with('[') || !ipv6_port_part.starts_with(':')) {
                            // malformed address, do not try to process it.
                            return false;
                    }
                    addressPart = std::string(ipv6_address_part.substr(1)); // drop '['
                    portPart    = std::string(ipv6_port_part   .substr(1)); // drop ':'
                    setIPv6();
            } else {
                    auto numColons = ranges::count(address, ':');
                    if (numColons == 0) {
                            // either IPv4 or IPv6
                            addressPart = std::string(address);
                    } else if (numColons == 1) {
                            // ipv4:port
                            auto [ipv4_address_part, ipv4_port_part] = StringOp::splitOnFirst(address, ':');
                            addressPart = std::string(ipv4_address_part);
                            portPart    = std::string(ipv4_port_part);
                            setIPv4();
                    } else {
                            // ipv6 address
                            addressPart = std::string(address);
                            setIPv6();
                    }
            }
    
            // Interpret 'addressPart' and 'portPart' (possibly the latter is empty)
            struct addrinfo* res;
            if (getaddrinfo(addressPart.c_str(), portPart.c_str(), &hints, &res) != 0) {
                    return false;
            }
    
            memcpy(&socket_address.address, res->ai_addr, res->ai_addrlen);
            freeaddrinfo(res);
            return true;
    }

I hope I didn't make any mistakes while rewriting this code.

Note 1: this requires these #includes at the top of the .cc file:

    #include "ranges.hh"
    #include "StringOp.hh"

Note 2: I believe calling `freeaddrinfo()` is only needed when getaddrinfo() returns success. At least that would be logical, but I'm not 100% sure about it (and the documentation isn't clear).